Remove curves that do not exist any more when refreshing a plot

DannyHindson opened the pull request:
Resolves error in Workbench when running algorithms that reduce the number
of spectra in a workspace (eg SumSpectra) while having plot window open. The spectra that don't exist are also removed from the plot. The required behaviour already exists in MantidPlot.

**Description of work.**

Changed Workbench python code (Framework/PythonInterface/mantid/plots/helperfunctions.py) to catch the run time error raised in C++ MatrixWorkspace.cpp when the spectra previously shown on the plot are not found. A new exception is raised with a better message containing the offending spectrum number and this is caught in PythonInterface\mantid\plots\\_\_init__.py where the related curve is removed from the plot window and the legend

Note - not possible to debug the specific piece of python code changed in PyCharm for some reason because the thread of execution has gone from Python into C++ code and back out into Python again. Discussed with Martin who said it's a known issue.

**Report to**: James Lord

**To test:**

1. Load data file containing multiple spectra into Workbench eg HRP39182.raw
2. Plot multiple spectra eg spectra 1-10 and leave window open
3. Run SumSpectra algorithm
4. before fix a run time error is shown "Could not find spectrum number in any spectrum", after fix there is an information message indicating the missing curves were not found and have not been plotted. 

Note if you don't specify the spectra list in step 2 (eg just plot all) the error displayed before the fix is slightly different

Fixes #26996 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Remove curves that do not exist any more when refreshing a plot

Resolves error in Workbench when running algorithms that reduce the number
of spectra in a workspace (eg SumSpectra) while having plot window open

DannyHindson committed:
Add unit test and fix problem with calling remove_workspace_artists

DannyHindson committed:
Add is empty handling to replace_workspace_artists

Add is empty handling to replace_workspace_artists to fully clear up deleted curves

hsautessella commented on Framework/PythonInterface/mantid/plots/__init__.py:
```diff
@@ -605,11 +613,32 @@ def plot(self, *args, **kwargs):
+                        x, y, _, __ = plotfunctions._plot_impl(self, workspace, args, kwargs)
+                    artists[0].set_data(x, y)
+                except RuntimeError as ex:
+                    logger.information('Curve not plotted: {0}'.format(ex.args[0]))
```
Could we leave a comment here saying why we might expect an error?

DannyHindson committed:
Incorporate review comments

DannyHindson commented:
Have made changes to address all review points. Rather than adding an additional call to draw() I have converted an existing call to draw_idle() to a draw. The draw_idle() call wasn't doing anything because a draw was already pending

DannyHindson committed:
Missed a review comment about error bars. Change made now

DannyHindson committed:
Resolve conflict vs master on plots __init__.py

DannyHindson commented:
I've resolved a conflict on the mantid\plots __init__.py unit. Following the merge, the plot wasn't rescaling after running SumSpectra so I adjusted the logic from your change (Harry) in the _data_update function (~L620) a bit to align it with the behaviour for error bars. You'll see I now have an _autoscale variable same as in the other _data_update functions

I wondered also if this piece of code was correct that appears 2-3 times:

            if self.lines:
                # Otherwise set autoscale to autoscale_on_update.
                self.set_autoscaley_on(autoscale_on_update) <----

Should the parameter be simply autoscale (boolean) rather than autoscale_on_update (which is a function\context manager)?

DannyHindson committed:
Ensure plot rescales properly after SumSpectra run

hsautessella commented:
The `auto_scale_on_udpate` context manger was to replace those lines, so they definitely should not be there. A big oversight on my part forgetting to remove those. Would you remove both instances from the `_data_update` functions? 

We will need to double check that #27047 is still doing what it's supposed to.

Otherwise, I think this is looking good and working as expected.

DannyHindson committed:
Update old references to autoscale_on_update

autoscale_on_update has been redefined as a context manager. It used to be a boolean
There will still some old references to autoscale_on_update which expected it to be
a boolean so these have been updated to use an alternative variable name

DannyHindson commented:
I've updated these lines to refer to autoscale rather than autoscale_on_update (which is now the context manager):

if self.lines:
            # Otherwise set autoscale to autoscale_on_update.
            self.set_autoscaley_on(autoscale) <----

I didn't delete them in the end because I didn't see the logic around whether self.lines was empty in the new context manager. So seemed to be doing something useful

hsautessella commented on Framework/PythonInterface/mantid/plots/__init__.py:
```diff
@@ -631,7 +660,7 @@ def _data_update(artists, workspace, new_kwargs=None):
             if self.lines:
                 # Otherwise set autoscale to autoscale_on_update.
-                self.set_autoscaley_on(autoscale_on_update)
+                self.set_autoscaley_on(autoscale)
```
As discussed offline, the best thing to do here is probably incorporate the `if` statement into the context manger 👍

DannyHindson committed:
Move if self.lines inside autoscale context manager

mantid-builder commented:
@DannyHindson could you review the feedback left on this PR and make changes as appropriate
<!-- {"problem_type": "ignored_review"} -->

DannyHindson commented:
I have moved the self.lines logic into the context manager as discussed with Harry

DannyHindson commented:
@hsautessella - are you OK to finish off reviewing this PR? This is the one we had a chat about relating to the autoscale context mgr you put in. There was one v small change left which was to move a check on self.lines into the context mgr.

hsautessella commented:
I think @alicerussell1 is going to give this one last functional review. The code looks good to me

alicerussell1 reviewed (dismissed):
This looks good to me, fixes the errors

DannyHindson commented:
Thanks both @alicerussell1 @hsautessella . I think you need to approve it as well Harry before it can be merged

hsautessella commented:
It only needs to be approved by one person + a gatekeeper. But I'll approve anyway, 'cause why not!

DanNixon commented on Framework/PythonInterface/mantid/plots/__init__.py:
```diff
@@ -466,8 +467,15 @@ def replace_workspace_artists(self, workspace):
             return False
 
+        is_empty_list = []
         for workspace_artist in artist_info:
```
This loop could be replaced with a generator, e.g:
```python
is_empty_list = [workspace_artist.replace_data(workspace) for workspace_artist in artist_info]
```

DannyHindson replied:
I had copied this loop from two similar loops that also created an is_empty_list higher up in the same unit. I've changed all 3 instances now to use the list comprehension you suggested

DanNixon commented on Framework/PythonInterface/mantid/plots/__init__.py:
```diff
@@ -611,12 +619,33 @@ def plot(self, *args, **kwargs):
                 # It's only possible to plot 1 line at a time from a workspace
+                try:
+                    if new_kwargs:
+                        x, y, _, __ = plotfunctions._plot_impl(self, workspace, args, new_kwargs)
```
Does repeated use of `_` trigger a static analysis warning? If not it might be better to just use single underscores for throwaway variables (Python does allow this).

DannyHindson replied:
OK I've replaced the double underscore with a single underscore

DanNixon commented on Framework/PythonInterface/mantid/plots/__init__.py:
```diff
@@ -611,12 +619,33 @@ def plot(self, *args, **kwargs):
-                if new_kwargs and new_kwargs.pop('autoscale_on_update', self.get_autoscale_on()):
+                    _autoscale = self.get_autoscale_on()
+
+                if _autoscale:
```
```suggestion
                if autoscale_enabled:
```

Or something to indicate this controls if auto scaling happens.

DannyHindson replied:
OK I've renamed this variable to _autoscale_on which occurred in 2 places in this unit. I have also renamed a similar autoscale variable in the same unit

DanNixon reviewed (changes requested):
Looks good, just a couple of style comments.

DannyHindson committed:
Incorporate changes from Dan Nixon review

gemmaguest commented:
Does this need release notes?

Also, in future could you try to avoid generic commit messages like "Incorporate review comments" and be more specific about what the commit contains (like you have for the other commits :slightly_smiling_face:)

DannyHindson committed:
Add release notes for bug fix

Running SumSpectra on an active plot no longer causes an error

DannyHindson committed:
Merge branch 'master' into 26996_WorkbenchPlotWindowError

DannyHindson commented:
I've added a line to the workbench release notes to mention this bug fix.
Comment about commit messages noted and will be more verbose in future. Was assuming the review comments were visible on the PR but guess this isn't always to hand when looking at the git history

DannyHindson committed:
Merge branch 'master' into 26996_WorkbenchPlotWindowError

DannyHindson commented:
Hi @gemmaguest - please can you give this PR a final check\approval? The last change was an addition to the release notes which triggered a new conflict today as other people add content to that .rst file so would be good to get it closed off soon. Thanks

gemmaguest commented:
The reviews had just been dismissed by a change to the release notes so merged anyway.
